File: src/FluentCards/ImageSetBuilder.cs
namespace FluentCards;

/// <summary>
/// Fluent builder for creating ImageSet elements.
/// </summary>
public class ImageSetBuilder
{
    private readonly ImageSet _imageSet = new() { Images = new List<Image>() };

    /// <summary>
    /// Sets the unique identifier for the image set.
    /// </summary>
    /// <param name="id">The unique identifier.</param>
    /// <returns>The builder instance for method chaining.</returns>
    public ImageSetBuilder WithId(string id)
    {
        _imageSet.Id = id;
        return this;
    }

    /// <summary>
    /// Sets the size for all images in the set.
    /// </summary>
    /// <param name="size">The image size.</param>
    /// <returns>The builder instance for method chaining.</returns>
    public ImageSetBuilder WithImageSize(ImageSize size)
    {
        _imageSet.ImageSize = size;
        return this;
    }

    /// <summary>
    /// Adds an image to the image set.
    /// </summary>
    /// <param name="configure">Action to configure the image.</param>
    /// <returns>The builder instance for method chaining.</returns>
    public ImageSetBuilder AddImage(Action<ImageBuilder> configure)
    {
        var builder = new ImageBuilder();
        configure(builder);
        _imageSet.Images!.Add(builder.Build());
        return this;
    }

    /// <summary>
    /// Adds an image to the image set.
    /// </summary>
    /// <param name="image">The image to add.</param>
    /// <returns>The builder instance for method chaining.</returns>
    public ImageSetBuilder AddImage(Image image)
    {
        _imageSet.Images!.Add(image);
        return this;
    }

    /// <summary>
    /// Builds and returns the configured ImageSet.
    /// </summary>
    /// <returns>The configured ImageSet instance.</returns>
    public ImageSet Build()
    {
        return _imageSet;
    }
}

File: go/fluentcards/image_set_builder.go
package fluentcards

// ImageSetBuilder builds an ImageSet Adaptive Card element.
type ImageSetBuilder struct {
	data map[string]any
}

func newImageSetBuilder() *ImageSetBuilder {
	return &ImageSetBuilder{data: map[string]any{"type": "ImageSet", "images": []any{}}}
}

func (b *ImageSetBuilder) WithID(id string) *ImageSetBuilder {
	b.data["id"] = id
	return b
}

func (b *ImageSetBuilder) WithImageSize(size ImageSize) *ImageSetBuilder {
	b.data["imageSize"] = string(size)
	return b
}

func (b *ImageSetBuilder) WithSpacing(spacing Spacing) *ImageSetBuilder {
	b.data["spacing"] = string(spacing)
	return b
}

func (b *ImageSetBuilder) WithIsVisible(isVisible bool) *ImageSetBuilder {
	b.data["isVisible"] = isVisible
	return b
}

func (b *ImageSetBuilder) WithSeparator(separator bool) *ImageSetBuilder {
	b.data["separator"] = separator
	return b
}

func (b *ImageSetBuilder) WithHeight(height string) *ImageSetBuilder {
	b.data["height"] = height
	return b
}

func (b *ImageSetBuilder) WithFallback(fallback any) *ImageSetBuilder {
	b.data["fallback"] = fallback
	return b
}

func (b *ImageSetBuilder) WithRequires(key, version string) *ImageSetBuilder {
	reqs, ok := b.data["requires"].(map[string]any)
	if !ok {
		reqs = map[string]any{}
	}
	reqs[key] = version
	b.data["requires"] = reqs
	return b
}

func (b *ImageSetBuilder) WithRtl(rtl bool) *ImageSetBuilder {
	b.data["rtl"] = rtl
	return b
}

// AddImageadds an image configured by the provided function.
func (b *ImageSetBuilder) AddImage(configure func(*ImageBuilder)) *ImageSetBuilder {
	ib := newImageBuilder()
	configure(ib)
	images := b.data["images"].([]any)
	b.data["images"] = append(images, ib.Build())
	return b
}

func (b *ImageSetBuilder) Build() Card {
	return b.data
}
